tools/lua_sstable_consumer.cc

```diff
@@ -0,0 +1,1003 @@
+        lua_close(l);
+    }
+};
+
+void* lua_alloc(void* ud, void* ptr, size_t osize, size_t nsize) {
+    if (!nsize) {
+        free(ptr);
+        return nullptr;
+    } else {
+        return realloc(ptr, nsize);
+    }
+}
+
+static const luaL_Reg loadedlibs[] = {
+    {"base", luaopen_base},
+    {LUA_STRLIBNAME, luaopen_string},
+    {LUA_TABLIBNAME, luaopen_table},
+    {NULL, NULL},
+};
+
+template <typename T>
+void push_userdata_ref(lua_State* l, T& v, const char* metatable_name) {
+    *reinterpret_cast<T**>(lua_newuserdata(l, sizeof(T*))) = &v;
+    if (metatable_name) {
+        luaL_setmetatable(l, metatable_name);
+    }
+}
+
+template <typename T>
+T& pop_userdata_ref(lua_State* l, int i) {
+    return **reinterpret_cast<T**>(lua_touserdata(l, i));
+}
+
+template <typename T>
+void push_userdata(lua_State* l, T v, const char* metatable_name) {
+    new (lua_newuserdata(l, sizeof(T))) T(std::move(v));
+    if (metatable_name) {
+        luaL_setmetatable(l, metatable_name);
+    }
+}
+
+// Push userdata with default constructed T in it.
+template <typename T>
+void push_userdata(lua_State* l, const char* metatable_name) {
+    new (lua_newuserdata(l, sizeof(T))) T;
+    if (metatable_name) {
+        luaL_setmetatable(l, metatable_name);
+    }
+}
+
+template <typename T>
+T pop_userdata(lua_State* l, int i) {
+    return std::move(*reinterpret_cast<T*>(lua_touserdata(l, i)));
+}
+
+// Get reference to userdata stored by-value.
+// Valid only until Lua GC:s the value, so don't keep it around for long.
+template <typename T>
+T& pop_userdata_as_ref(lua_State* l, int i) {
+    return *reinterpret_cast<T*>(lua_touserdata(l, i));
+}
+
+template <typename T>
+int cxx_gc_l(lua_State* l) {
+    auto& v = pop_userdata_as_ref<T>(l, 1);
+    v.~T();
+    lua_pop(l, 2);
+    return 0;
+}
+
+const schema& get_schema_l(lua_State* l) {
+    lua_getglobal(l, "schema");
+    auto& s = pop_userdata_ref<const schema>(l, -1);
+    lua_pop(l, 1);
+    return s;
+}
+
+int sstable_index_l(lua_State* l) {
+    auto field = lua_tostring(l, 2);
+    auto& sst = pop_userdata_ref<const sstables::sstable>(l, 1);
+    lua_pop(l, 2);
+    if (!strcmp(field, "filename")) {
+        const auto f = sst.get_filename();
+        lua_pushstring(l, f.data());
+        return 1;
+    }
+    return 0;
+}
+
+int partition_start_index_l(lua_State* l) {
+    auto field = lua_tostring(l, 2);
+    auto& ps = pop_userdata_as_ref<partition_start>(l, 1);
+    lua_pop(l, 2);
+    auto& schema = get_schema_l(l);
+    if (!strcmp(field, "key")) {
+        const auto f = format("{}", ps.key().key().with_schema(schema));
+        lua_pushstring(l, f.data());
+        return 1;
+    } else if (!strcmp(field, "key_raw")) {
+        auto mb = managed_bytes(ps.key().key().representation());
+        push_userdata<managed_bytes>(l, std::move(mb), managed_bytes_metatable);
+        return 1;
+    } else if (!strcmp(field, "token")) {
+        lua_pushinteger(l, ps.key().token()._data);
```

Comment:
This will break: https://www.lua.org/pil/2.3.html

Although, to_lua_visitor also uses lua_pushinteger. Maybe it auto-expands to varint?